### ISSUE TITLE
Premature thread exit (possibly uwsgi's fault)

### ISSUE THREAD

**Issue by b0urb0n** [dchevell/flask-executor#9]
I'm using `uwsgi` to run my Flask app and I have a need to spawn threads to do some work that takes between 10 seconds and several minutes.  I have a database to manage the state of the tasks.  On a POST, I need to act on what was POSTed.  I'm using `Flask-RESTful`, `Flask-SQLAlchemy`, and I'm trying to use `Flask-Executor`

I define the executor in the main app:
```
executor = Executor()
def create_app(cfg) -> Flask:
    ...
    executor.init_app(app)
    ...
```

I call the worker during POST and return the serialized task which can be used to come back and get the status or the results.
```
class Base(Resource):
    def post(self):
        # handle incoming POST data
        logger.info("POST (create) task for {}".format(data.id))

        task = models.new_task(data.id)  # new Task ORM
        my_worker.submit(task)

        return task.serialize()
```

The worker
```
@executor.job
def my_worker(task):
    logger.info("Starting task {}".format(task.id))
    models.update_status(task, models.STATUS_IN_PROGRESS)

    # do work

    logger.info("Writing report for task {}".format(task.id))
    # write results to a file

    models.update_status(task, models.STATUS_COMPLETE)
    logger.info("Task {} complete".format(task.id))
```

I think my issue is the fault of `uwsgi`.  I believe the thread/process that spawned my worker thread exits before my worker thread finishes its work.  I think this because I see the very first log entry almost every time (`Starting task X`) and then nothing follows.

Have you encountered this issue at all?  If so, how did you solve it?  It may have something to do with this SO post: https://stackoverflow.com/q/49992329/1694291

**Comment by dchevell** [comment]
Hey @b0urb0n!

tl;dr did you try running uWSGI with `--enable-threads`?

Long version: I tested this and reproduced the issue like so:

```python
# test.py

from time import sleep

from flask import Flask
from flask_executor import Executor

app = Flask(__name__)
executor = Executor(app)

def longtask():
    print('before task')
    sleep(5)
    print('after task')

@app.route('/task')
def task():
    executor.submit(longtask)
    sleep(1)
    return 'ok'
```
I ran uWSGI like so:
```bash
$ uwsgi --http :9090 --wsgi-file test.py --callable app
```

Sure enough, when I hit my test endpoint the "before task" line prints, but the "after task" line does not. So far, sounds exactly like what you described. I did a little digging and found [this note](https://uwsgi-docs.readthedocs.io/en/latest/ThingsToKnow.html) in the uWSGI docs: 

> By default the Python plugin does not initialize the GIL. This means your app-generated threads will not run. If you need threads, remember to enable them with `enable-threads`. Running uWSGI in multithreading mode (with the `threads` options) will automatically enable threading support. This “strange” default behaviour is for performance reasons, no shame in that.

I tried again with `--enable-threads`:
```bash
$ uwsgi --http :9090 --wsgi-file test.py --callable app --enable-threads
```

Voila, now my background task completes and prints both messages. By the sounds of it this is expected behaviour which is overcome by running uWSGI in multithreading mode.

Hope that helps! Let me know how you go.

**Comment by b0urb0n** [comment]
Well, that seems to have done the trick.  I had no idea that's what my problem was.  Now to track down the thread can't seem to update my database that the task is complete.

Thanks!
